src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -0,0 +1,151 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Strategies;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    public static partial class RandomAccess
+    {
+        // IovStackThreshold matches Linux's UIO_FASTIOV, which is the number of 'struct iovec'
+        // that get stackalloced in the Linux kernel.
+        private const int IovStackThreshold = 8;
+
+        internal static long GetFileLength(SafeFileHandle handle, string? path)
+        {
+            int result = Interop.Sys.FStat(handle, out Interop.Sys.FileStatus status);
+            FileStreamHelpers.CheckFileCall(result, path);
+            return status.Size;
+        }
+
+        private static unsafe int ReadAtOffset(SafeFileHandle handle, Span<byte> buffer, long fileOffset)
+        {
+            fixed (byte* bufPtr = &MemoryMarshal.GetReference(buffer))
+            {
+                int result = Interop.Sys.PRead(handle, bufPtr, buffer.Length, fileOffset);
+                FileStreamHelpers.CheckFileCall(result, path: null);
+                return result;
+            }
+        }
+
+        private static unsafe long ReadScatterAtOffset(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset)
+        {
+            MemoryHandle[] handles = new MemoryHandle[buffers.Count];
+            Span<Interop.Sys.IOVector> vectors = buffers.Count <= IovStackThreshold ? stackalloc Interop.Sys.IOVector[IovStackThreshold] : new Interop.Sys.IOVector[buffers.Count];
+
+            long result;
+            try
+            {
+                for (int i = 0; i < buffers.Count; i++)
```

Comment:
buffers is an interface.  If you cache buffers.Count into a local and then iterate with that, you'll avoid N-1 interface dispatches.  Same for the other cases of it in the rest of the PR.